File: 2021년/7월/0709 맞춰봐.cpp
#include <iostream>

using namespace std;

int n, s[10][10], nums[10];
char lst[10][10];

bool check(int index)
{
	int sum = 0;
	for(int i=index;i>-1;i--)
	{
		sum += nums[i];
		if (s[i][index] == 0 && sum != 0) return false;
		if (s[i][index] == 1 && sum <= 0) return false;
		if (s[i][index] == -1 && sum >= 0) return false;
	}
	return true;
}

void solve(int index)
{
	if (index == n)
	{
		for(int i=0;i<n;i++) cout << nums[i] << " ";
		exit(0);
	}
	
	if (s[index][index] == 0)
	{
		nums[index] = 0;
		solve(index+1);
		return;
	}
	
	for(int i=1;i<11;i++)
	{
		nums[index] = i*s[index][index];
		if (check(index) == true) solve(index+1);
	}
}

int main()
{
	cin >> n;
	for(int i=0;i<n;i++)
	{
		for(int j=i;j<n;j++)
		{
			cin >> lst[i][j];
			if(lst[i][j] == '+') s[i][j] = 1;
			else if(lst[i][j] == '-') s[i][j] = -1;
			else s[i][j] = 0;
		}
	}
	
	solve(0);
}

File: 2021년/7월/0709 맞춰봐.py
# 백준_1248_맞춰봐_백트래킹_골드 3

def solve(depth,num):
    if depth == n:

    
    if depth == 0:
        if s[0] == '+':
            nums.append()        
        elif s[0] == '-':
        
        elif s[0] == '0':

    for i in range(-10,10,1):
        if check(i,depth,num)
            nums.append(i)
            solve(depth+1,num)
            nums.pop()

def check(depth, i):
    if depth = 0:

    sum = 0
    for k in range(depth+1+num):
        sum += nums[k]



    for i in range(n):
        sum = 0
        for j in range(i,n):
            if i == j:
                if nums[i] > 0:
                    if s[] == '+':
                        return True
                    else:
                        return False

                elif nums[i] < 0:
                    if s[] == '-':
                        return True
                    else:
                        return False

                elif nums[i] == 0:
                    if s[] == '0':
                        return True
                    else:
                        return False
                continue

            sum += nums[i] + nums[j]
            if sum > 0:
                if s[] == '+':
                    return True
                else:
                    return False

            elif sum < 0:
                if s[] == '-':
                    return True
                else:
                    return False

            elif sum == 0:
                if s[] == '0':
                    return True
                else:
                    return False



n = int(input())
lst = list(input())
s = []
num_list = 0

for i in range(n,0,-1):
    s.append(lst[num_list:num_list+i])
    num_list += i

nums = []

solve(0,num)